### PR TITLE
Indirect - IqtFit - Display ties added by 'Constrain Intensities' in the Fit Property Browser

### LINKED ISSUE #21807: Indirect - IqtFit - Display ties created with 'Constrain Intensities' in fit property browser
Currently, in the IqtFit interface, the ties created by checking the check-box labelled 'Constrain Intensities' are not displayed within the fit property browser.

These ties should be made viewable within the fit property browser.

### PR DESCRIPTION
**Pull request by brandonhewer**
Currently, in the Indirect IqtFit interface, the tie created by the 'Constrain Intensities' option does not appear in the browser (although it is used in the fit).

This PR ensures the fit property browser displays the tie.

**Must be merged after PR #21962**

**To test:**
1. Navigate to Interfaces > Indirect > Data Analysis > IqtFit.
2. Add two exponential functions using the 'Exponential' spinner and a stretched exponential using the 'Stretched Exponential' checkbox.
3. Select 'Flat Background' from the 'Background Type' drop-down menu.
4. Check the check-box labelled 'Constrain Intensities'.
5. Ensure the tie can be viewed in the function browser found under the 'Functions' heading. The tie should be each of the Height parameters of the exponential functions and the A0 parameter of the background function, all subtracted from 1.
6. Uncheck the check-box labelled 'Constrain Intensities'.
7. Ensure the tie is removed.

Fixes #21807. 

**Release Notes** 

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [x] Do changes function as described? Add comments below that describe the tests performed?
- [x] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/scientific_interfaces/Indirect/IndirectDataAnalysisTab.cpp [hunk 1]
```diff
@@ -99,7 +95,9 @@ void IndirectDataAnalysisTab::setPreviewPlotWorkspace(
  *
  * @return  The selected spectrum.
  */
-int IndirectDataAnalysisTab::selectedSpectrum() { return m_selectedSpectrum; }
+int IndirectDataAnalysisTab::selectedSpectrum() const {
```
Comment (reimundILL) [inline qt/scientific_interfaces/Indirect/IndirectDataAnalysisTab.cpp]: In my opinion making those methods const is a really good idea.

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/IqtFitMultiple.py [hunk 2]
```diff
@@ -64,6 +64,11 @@ def PyInit(self):
         self.declareProperty(name='ConstrainIntensities', defaultValue=False,
                              doc="If the Intensities should be constrained during the fit")
 
+        self.declareProperty(name='ExtractMembers', defaultValue=False,
+                             doc="If true, then each member of the fit will be extracted, into their"
```
Comment (reimundILL) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/IqtFitMultiple.py]: Whitespace missing `theirown` ?

File: Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/IqtFitMultiple.py [hunk 3]
```diff
@@ -64,6 +64,11 @@ def PyInit(self):
...
+                                 "own workspace. These workspaces will have a histogram for each spectrum "
+                                 "(Q-value) and will be grouped.")
```
Comment (reimundILL) [inline Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/IqtFitMultiple.py]: I cannot read the rest of this tool-tip documentation. (Ubuntu 16.04).
Maybe too many `"`s ?

**Comment by reimundILL** [comment]
A remark on `IqtFitMultiple` (is not part of this issue): found some redundant lines when calling `CropWorkspace` (ll 142-158). From my point of view it would be enough to set the property `EndWorkspaceIndex` only if `_spec_max` is not None. All other properties seem to not change.

File: docs/source/diagrams/IqtFitMultiple-v1_wkflw.dot [hunk 4]
```diff
@@ -71,4 +71,7 @@ digraph IqtFitMultiple {
   SampleLogs                    -> AddSampleLogMultiple         [label="Additional SampleLogs based on Fit data"]
   AddSampleLogMultiple          -> ResultWorkspace2             [label="Outputworkspace"]
   AddSampleLogMultiple2         -> FitGroup2                    [label="OutputWorkspace"]
+  ExtractMembers                -> ExtractQENSMembers           [label="Extract members when true"]
```
Comment (reimundILL) [inline docs/source/diagrams/IqtFitMultiple-v1_wkflw.dot]: Very good to update the diagrams.

A suggestion first: would it be easier and quicker to read "Extract members ?"

In `IqtFitSequential`, the colors are different (yellow and blue), here they are gray. I think it should be the same for both algorithms. I guess `IqtFitSequential` has the desired look of this part of the diagram.

**Comment by brandonhewer** [comment]
@reimundILL The appropriate changes regarding documentation and the workflow diagrams have now been updated in the prerequisite PR (#21724). These changes have been merged into this PR.

**Comment by reimundILL** [comment]
Cool, I see. I will proceed testing ... will give a feedback tomorrow then. Have a good impression about the changes made.

**Comment by reimundILL** [comment]
I could crash MantidPlot when playing with  Interfaces > Indirect > Data Analysis > IqtFit. I tried foolowing your test instructions and clicked `Sequential` if I remember right. I see:

``` Python
MantidPlot: /usr/include/boost/smart_ptr/shared_ptr.hpp:648: typename boost::detail::sp_member_access<T>::type boost::shared_ptr<T>::operator->() const [with T = Mantid::API::MatrixWorkspace; typename boost::detail::sp_member_access<T>::type = Mantid::API::MatrixWorkspace*]: Assertion `px != 0' failed.
Aborted
```

Try now to reproduce the behavior & document in a better way.

**Comment by brandonhewer** [comment]
@reimundILL This appears to be new 😞, thanks for the find.

**Comment by reimundILL** [comment]
- Crash fixed

- Viewing tie:

**Comment by SpencerHowells** [comment]
@brandonhewer @reimundILL I have a problem with the tie for 2 functions. The background is the more reliable fit so the exp+strexp height is 1-bgd. For 1 function it is straight forward, but for 2 I cannot see how it can be done. So I just tie the first peak and let the second be free.

**Comment by brandonhewer** [comment]
@SpencerHowells So for 2 or more functions (not including StretchExp + ExpDecay), it is better to allow remaining functions to be free, such that only ```the first height = 1-Background```
Rather than ```the sum of all function heights = 1-Background```

**Comment by SpencerHowells** [comment]
@brandonhewer For 2 or more peaks the sum of heights = 1-bgd including exp + strexp. I couldn't see how to separate the exp + strexp (or exp1 + exp2) heights in the ties. ie do (exp_ht + strexp_ht) = 1 - bgd.

**Comment by brandonhewer** [comment]
@SpencerHowells Currently the tie for heights is set such that:
```ExpDecay = 1-StretchExp-Background```
or more generally:
```ExpDecay = 1 - (Sum of all other function heights) - Background```

**Comment by brandonhewer** [comment]
@reimundILL When you're available, would you be able to take another look at this PR? @SpencerHowells and I have discussed the above and it appears the current solution for intensity ties is ok.

**Comment by reimundILL** [comment]
@brandonhewer Thanks, will have a look now.

**Comment by reimundILL** [comment]
When playing with the interface I see:

screenshot from 2018-02-28 13-19-45

I followed your instructions, it looked nice. I changed Exponential, it happened. Now it seems to be stable again.

**Comment by reimundILL** [comment]
Fit Function -> Fit -> Evaluate Function is ok, input validated
Fit Function -> Sequential Fit shows the image mentioned above

**Comment by reimundILL** [comment]
Some suspicious terminal info:
```Python
MantidApplication-[Fatal] Unexpected exception: External function parameter name must start with 'f'
MantidApplication-[Fatal] Continue working.
MantidApplication-[Fatal] Unexpected exception: External function parameter name must start with 'f'
MantidApplication-[Fatal] Unexpected exception: External function parameter name must start with 'f'
MantidApplication-[Fatal] Unexpected exception: External function parameter name must start with 'f'
MantidApplication-[Fatal] Unexpected exception: External function parameter name must start with 'f'
```

**Comment by brandonhewer** [comment]
@reimundILL @SpencerHowells I've found the bug to be associated to the FitPropertyBrowser itself, which has revealed itself at this level due to an attempted clone of a CompositeFunction with multiple ties for the same parameter.

An issue for this has been created (#21919) and assigned to @mantid-roman:
https://github.com/mantidproject/mantid/issues/21919

This should hopefully fix the bug that has been found here.

**Comment by brandonhewer** [comment]
@reimundILL A recent fix by roman (not yet merged into the release) has been merged into this PR and appears to have solved the issue. This PR must now be merged after #21962, but should be available for testing.

**Comment by reimundILL** [comment]
Very good!

On master
- MantidPlot crashed at least one time when setting parameters
- Ties were not updated according to user option

Now
- Stable and robust GUI
- Setting1 (Exponential 1, Stretched Exponential True, no background, constrain intensities true)

- Setting2 (Exponential 1, Stretched Exponential True, flat background, constrain intensities true)

- Setting3 (Exponential 2, Stretched Exponential True, no background, constrain intensities true)

- Setting4 (Exponential 2, Stretched Exponential True, flat background, constrain intensities true)

**Comment by reimundILL** [comment]
@brandonhewer  do you consider to add release notes? One minor point: the tool-tip repeats the "value" and thus, is not very informative.

**Comment by brandonhewer** [comment]
@reimundILL Thanks for testing this PR throughout the issues here. 
The release notes were updated in: https://github.com/mantidproject/mantid/pull/21897

**Comment by reimundILL** [comment]
You're right, I remember.

**Review by vreimund — APPROVED** [review]

**Pull request merged**